### PR TITLE
Promoteds are statics and statics have a place, not just a value

### PR DESCRIPTION
**Pull request by oli-obk**
r? @eddyb 

This makes everything around promoteds a little simpler

### CODE REVIEW

File: src/librustc_codegen_llvm/mir/block.rs [hunk 1]
```diff
@@ -507,14 +507,37 @@ impl<'a, 'tcx> FunctionCx<'a, 'tcx> {
...
                             match *arg {
+                                mir::Operand::Copy(mir::Place::Promoted(box(index, ty))) |
+                                mir::Operand::Move(mir::Place::Promoted(box(index, ty))) => {
+                                    let param_env = ty::ParamEnv::reveal_all();
+                                    let cid = mir::interpret::GlobalId {
+                                        instance: self.instance,
+                                        promoted: Some(index),
+                                    };
+                                    let c = bx.tcx().const_eval(param_env.and(cid));
+                                    let (llval, ty) = self.simd_shuffle_indices(
```
Comment (nagisa) [inline src/librustc_codegen_llvm/mir/block.rs]: What is the point of this?

**Comment by nagisa** [comment]
This is r=me once https://github.com/rust-lang/rust/pull/52571 lands and the issues are fixed.

File: src/librustc_codegen_llvm/mir/block.rs (hunk 1 above)
Comment (nagisa) [reply src/librustc_codegen_llvm/mir/block.rs]: I see. A comment to that effect would be great.

File: src/librustc_codegen_llvm/mir/operand.rs [hunk 2]
```diff
@@ -409,20 +398,12 @@ impl<'a, 'tcx> FunctionCx<'a, 'tcx> {
...
-                                    "could not evaluate constant operand",
-                                );
-                            },
-                        }
+                        err.report_as_error(
+                            bx.tcx().at(constant.span),
+                            "could not evaluate constant operand",
+                        );
+                        // Allow RalfJ to sleep soundly knowing that even refactorings that remove
+                        // the above error (or silence it under some conditions) will not cause UB
```
Comment (RalfJung) [inline src/librustc_codegen_llvm/mir/operand.rs]: :heart:

**Comment by kennytm** [comment]
```
[01:30:33] error[E0004]: non-exhaustive patterns: `Promoted(_)` not covered
[01:30:33]    --> tools/miri/src/validation.rs:138:18
[01:30:33]     |
[01:30:33] 138 |         Ok(match *place {
[01:30:33]     |                  ^^^^^^ pattern `Promoted(_)` not covered
```

File: src/librustc_codegen_llvm/mir/block.rs [hunk 3]
```diff
@@ -507,14 +507,40 @@ impl<'a, 'tcx> FunctionCx<'a, 'tcx> {
                         // promotes any complex rvalues to constants.
                         if i == 2 && intrinsic.unwrap().starts_with("simd_shuffle") {
                             match *arg {
+                                // The shuffle array argument is usually not an explicit constant,
+                                // but specified directly in the code. This means it gets promoted
+                                // and we can then extract the value by evaluating the promoted.
+                                mir::Operand::Copy(mir::Place::Promoted(box(index, ty))) |
+                                mir::Operand::Move(mir::Place::Promoted(box(index, ty))) => {
+                                    let param_env = ty::ParamEnv::reveal_all();
```
Comment (eddyb) [inline src/librustc_codegen_llvm/mir/block.rs]: This should probably be something like `self.param_env` (minor future hazard if we start deduplicating instances).

File: src/librustc_mir/borrow_check/nll/type_check/mod.rs [hunk 4]
```diff
@@ -281,69 +281,48 @@ impl<'a, 'b, 'gcx, 'tcx> TypeVerifier<'a, 'b, 'gcx, 'tcx> {
...
+        // Trait<'a, 'b>>::SOME_CONST`) that impose
+        // constraints on `'a` and `'b`. These constraints
+        // would be lost if we just look at the normalized
+        // value.
+        if let ty::TyFnDef(def_id, substs) = constant.literal.ty.sty {
+            let tcx = self.tcx();
+            let type_checker = &mut self.cx;
+
+            // FIXME -- For now, use the substitutions from
+            // `value.ty` rather than `value.val`. The
```
Comment (eddyb) [inline src/librustc_mir/borrow_check/nll/type_check/mod.rs]: Heh, `value.val` mention here is outdated (cc @nikomatsakis).

File: src/librustc_codegen_llvm/mir/block.rs (hunk 3 above)
Comment (RalfJung) [reply src/librustc_codegen_llvm/mir/block.rs]: The PR already landed.Does anyone take care of fixing this?^^

Comment (oli-obk) [reply src/librustc_codegen_llvm/mir/block.rs]: This is rustc_codegen_llvm... we don't have a param env, everything is monomorphized

**Pull request merged**